test/parallel/test-domain-top-level-error-handler-clears-stack.js

```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+const assert = require('assert');
+const domain = require('domain');
+
+/*
+ * Make sure that the domains stack is cleared after a top-level domain
+ * error handler exited gracefully.
+ */
+const d = domain.create();
+
+d.on('error', function() {
+  process.nextTick(function() {
+    // Scheduling a callback with process.nextTick will enter a _new_ domain,
+    // and the callback will be called after the domain that handled the error
+    // was exited. So there should be only one domain on the domains stack if
+    // the domains stack was cleared properly when the domain error handler
+    // returned.
+    if (domain._stack.length !== 1) {
+      // Do not use assert to perform this test: this callback runs in a
+      // different callstack as the original process._fatalException that
+      // handled the original error, thus throwing here would trigger another
+      // call to process._fatalException, and son on recursively and
+      // indefinitely.
+      process.exit(1);
```

Comment:
This code path runs when the test already fails. However, using `assert` would, as indicated in the comment above, make the test time out. Using `process.exit` allows the test to complete right away.

The following output illustrates that:
```
➜  node git:(port-domains-clears-stack-test) ✗ git diff
diff --git a/lib/domain.js b/lib/domain.js
index b6321a2..eda8409 100644
--- a/lib/domain.js
+++ b/lib/domain.js
@@ -67,8 +67,8 @@ Domain.prototype._errorHandler = function errorHandler(er) {
     // Exit all domains on the stack.  Uncaught exceptions end the
     // current tick and no domains should be left on the stack
     // between ticks.
-    stack.length = 0;
-    exports.active = process.domain = null;
+    //stack.length = 0;
+    //exports.active = process.domain = null;
 
     return handled;
   }
diff --git a/test/parallel/test-domain-top-level-error-handler-clears-stack.js b/test/parallel/test-domain-top-level-error-handler-clears-stack.js
index 39eb861..7258199 100644
--- a/test/parallel/test-domain-top-level-error-handler-clears-stack.js
+++ b/test/parallel/test-domain-top-level-error-handler-clears-stack.js
@@ -3,6 +3,10 @@
 const domain = require('domain');
 const common = require('../common');
 
+setTimeout(function onTimeout() {
+  console.log('timed out');
+  process.exit(1);
+}, 1000);
 /*
  * Make sure that the domains stack is cleared after a top-level domain
  * error handler exited gracefully.
@@ -22,7 +26,7 @@ d.on('error', function() {
       // handled the original error, thus throwing here would trigger another
       // call to process._fatalException, and son on recursively and
       // indefinitely.
-      process.exit(1);
+      assert.fail('boom');
     }
   });
 });
➜  node git:(port-domains-clears-stack-test) ✗ make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C out BUILDTYPE=Release V=1
make[1]: Nothing to be done for `all'.
ln -fs out/Release/node node
➜  node git:(port-domains-clears-stack-test) ✗ ./node test/parallel/test-domain-top-level-error-handler-clears-stack.js
timed out
➜  node git:(port-domains-clears-stack-test) ✗ echo $?
1
➜  node git:(port-domains-clears-stack-test) ✗ git diff                                                                
diff --git a/lib/domain.js b/lib/domain.js
index b6321a2..eda8409 100644
--- a/lib/domain.js
+++ b/lib/domain.js
@@ -67,8 +67,8 @@ Domain.prototype._errorHandler = function errorHandler(er) {
     // Exit all domains on the stack.  Uncaught exceptions end the
     // current tick and no domains should be left on the stack
     // between ticks.
-    stack.length = 0;
-    exports.active = process.domain = null;
+    //stack.length = 0;
+    //exports.active = process.domain = null;
 
     return handled;
   }
diff --git a/test/parallel/test-domain-top-level-error-handler-clears-stack.js b/test/parallel/test-domain-top-level-error-handler-clears-stack.js
index 39eb861..ffe9137 100644
--- a/test/parallel/test-domain-top-level-error-handler-clears-stack.js
+++ b/test/parallel/test-domain-top-level-error-handler-clears-stack.js
@@ -3,6 +3,10 @@
 const domain = require('domain');
 const common = require('../common');
 
+setTimeout(function onTimeout() {
+  console.log('timed out');
+  process.exit(1);
+}, 1000);
 /*
  * Make sure that the domains stack is cleared after a top-level domain
  * error handler exited gracefully.
➜  node git:(port-domains-clears-stack-test) ✗ ./node test/parallel/test-domain-top-level-error-handler-clears-stack.js
➜  node git:(port-domains-clears-stack-test) ✗ echo $?
1
➜  node git:(port-domains-clears-stack-test) ✗
```